interop/grpclb_fallback/client.go

```diff
@@ -0,0 +1,217 @@
+        fast_fallback_after_startup : LB/backend connections fail fast after RPC's have been made;
+        slow_fallback_before_startup : LB/backend connections black hole before RPC's have been made;
+        slow_fallback_after_startup : LB/backend connections black hole after RPC's have been made;`)
+)
+
+type rpcMode uint8
+
+const (
+	// per-rpc settings
+	failFast     rpcMode = iota
+	waitForReady         = iota
+)
+
+func doRPCAndGetPath(client testpb.TestServiceClient, deadlineSeconds int, mode rpcMode) testpb.GrpclbRouteType {
+	grpclog.Infof("doRPCAndGetPath deadlineSeconds:%v rpcMode:%v", deadlineSeconds, mode)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(deadlineSeconds)*time.Second)
+	defer cancel()
+	copts := []grpc.CallOption{}
+	if mode == waitForReady {
+		copts = append(copts, grpc.WaitForReady(true))
+	}
+	req := &testpb.SimpleRequest{
+		FillGrpclbRouteType: true,
+	}
+	reply, err := client.UnaryCall(ctx, req, copts...)
+	if err != nil {
+		grpclog.Infof("doRPCAndGetPath error:%v", err)
+		return testpb.GrpclbRouteType_GRPCLB_ROUTE_TYPE_UNKNOWN
+	}
+	g := reply.GetGrpclbRouteType()
+	grpclog.Infof("doRPCAndGetPath got grpclb route type: %v", g)
+	if g != testpb.GrpclbRouteType_GRPCLB_ROUTE_TYPE_FALLBACK && g != testpb.GrpclbRouteType_GRPCLB_ROUTE_TYPE_BACKEND {
+		grpclog.Fatalf("Expected grpclb route type to be either backend or fallback", g)
+	}
+	return g
+}
+
+func dialTCPUserTimeout(ctx context.Context, addr string) (net.Conn, error) {
+	control := func(network, address string, c syscall.RawConn) error {
+		var syscallErr error
+		controlErr := c.Control(func(fd uintptr) {
+			syscallErr = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, unix.TCP_USER_TIMEOUT, 20000)
+		})
+		if syscallErr != nil {
+			grpclog.Fatalf("syscall error setting sockopt TCP_USER_TIMEOUT: %v", syscallErr)
+		}
+		if controlErr != nil {
+			grpclog.Fatalf("control error setting sockopt TCP_USER_TIMEOUT: %v", syscallErr)
+		}
+		return nil
+	}
+	d := &net.Dialer{
+		Control: control,
+	}
+	return d.DialContext(ctx, "tcp", addr)
+}
+
+func createFallbackTestConn() *grpc.ClientConn {
+	opts := []grpc.DialOption{
+		grpc.WithContextDialer(dialTCPUserTimeout),
+		grpc.WithBlock(),
+	}
+	switch *customCredentialsType {
+	case "tls":
+		creds := credentials.NewClientTLSFromCert(nil, "")
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	case "alts":
+		creds := alts.NewClientCreds(alts.DefaultClientOptions())
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	case "google_default_credentials":
+		opts = append(opts, grpc.WithCredentialsBundle(google.NewDefaultCredentials()))
+	case "compute_engine_channel_creds":
+		opts = append(opts, grpc.WithCredentialsBundle(google.NewComputeEngineCredentials()))
+	default:
+		grpclog.Fatalf("Invalid --custom_credentials_type:%v", *customCredentialsType)
+	}
+	conn, err := grpc.Dial(*serverURI, opts...)
```

Comment:
Since this is blocking, do a DialContext with timeout.

Reply:
Sorry for delay on this PR. This comment made me realize that this test is dramatically different for Go than it is for Java and C++ as written, so taking some time to take a closer look at things and revise